src/Apps/CheezeWizards/Services/Firebase/FirebaseService.js

```diff
@@ -0,0 +1,239 @@
+            .then(snapshots => {
+                if (snapshots.empty) {
+                    return [];
+                }
+                const challenges = [];
+                snapshots.docs.forEach(doc => {
+                    challenges.push({challengeId: doc.id, ...doc.data()});
+                });
+                return challenges;
+            });
+    }
+
+    // Todo: should handle network not being defined or invalid + associated test
+    // Todo: should handle wizards being undefined / not an array as the map will fail + associated test
+    // Todo: should guard against wizard.id being undefined + associated test
+    async upsertWizards(network, wizards) {
+
+        checkValidNetwork(network);
+
+        return Promise.all(wizards.map((wizard) => {
+            // /wizards/network/{networkID}/{wizardID}/
+            return db
+                .collection('wizards')
+                .doc('network')
+                .collection(network)
+                .doc(wizard.id)
+                .set(wizard, {
+                    merge: true
+                });
+        }));
+    }
+
+    // Todo: should handle network not being defined or invalid + associated test
+    // Todo: should check params passed in valid + associated test
+    // Todo: do we need to do ownership check that we're challenging someone else's wizard?
+    async sendChallenge(network, {challengeId, challengingWizardId, otherWizardId}) {
+
+        checkValidNetwork(network);
+
+        // challenger data
+        const challengerData = {
+            currentDuel: '0x0000000000000000000000000000000000000000000000000000000000000000',
+            otherWizardId,
+            challengeAccepted: false,
+            challenger: true,
+        };
+
+        // challengee data
+        const challengeeData = {
+            currentDuel: '0x0000000000000000000000000000000000000000000000000000000000000000',
+            challengingWizardId,
+            commitmentHash: '',
+            challengeAccepted: false,
+            challenger: false,
+        };
+
+        const networkRef = db.collection('wizards').doc('network').collection(network);
+        const challengerDataRef = networkRef.doc(challengingWizardId).collection('duel').doc(challengeId);
+        const challengeeDataRef = networkRef.doc(otherWizardId).collection('duel').doc(challengeId);
+
+        await db.runTransaction(t => {
+            t.set(challengerDataRef, challengerData);
+            t.set(challengeeDataRef, challengeeData);
+            return Promise.resolve('done');
+        });
+    }
+
+    async acceptChallenge(network, {wizardId, challengeId, commitmentHash, currentDuel}) {
+
+        checkValidNetwork(network);
+
+        const networkRef = db.collection('wizards').doc('network').collection(network);
+        const challengeRef = networkRef
+            .doc(wizardId)
+            .collection('duel')
+            .doc(challengeId);
+
+        await db.runTransaction(t => {
```

Comment:
Nice work of transactionality boundaries 👍

Reply:
Yeah it was great to figure this out